### ISSUE TITLE
[Bug] bf16 MLA decode still wraps past a 4 GiB window from `ptr_KV` on gfx950 — qh16 / qh32 / qh128 (qh64 is fixed)

### ISSUE THREAD

**Issue by afriedri** [ROCm/aiter#5765]
## Summary

`aiter.mla.mla_decode_fwd` addresses the KV cache through a 32-bit offset from
`ptr_KV`. Once a KV row lies more than `2**32` bytes past the base pointer the
kernel reads the wrong memory and returns approximately zero. There is no
fault, no NaN, and no warning — it surfaces only as a silent accuracy collapse
in whatever is using it.

This is the same class of defect as #2768, and PR #4156 fixed it for the
**qh64** bf16 decode kernels by moving them to a 64-bit `global_load_lds` path.
Sweeping `gqa_ratio` on current `amd-aiter 0.1.21.post2` shows the other bf16
decode variants were not carried along:

| `gqa_ratio` | kernel selected | > 4 GiB from `ptr_KV` |
|---|---|---|
| 16 | `mla_dec_stage1_bf16_a16w16_subQ16_mqa16.co` | **wraps** |
| 32 | `MLA_A16W16_1TG_4W_16mx4_32nx1_Coex0_Msk1_QH16.co` | **wraps** |
| 64 | `mla_a16w16_qh64_qseqlen1_gqaratio64_v3.co` | OK (PR #4156) |
| 128 | `mla_dec_stage1_bf16_a16w16_subQ128_mqa128.co` | **wraps** |

It affects both sparse decode (`page_size=1`) and ordinary paged decode
(`page_size=16`), so it is not specific to the sparse-MLA index path.

## Environment

- **GPU**: MI350X, gfx950 (`gfx950:sramecc+:xnack-`), 8×
- **ROCm**: 7.2.3
- **torch**: 2.12.0+git6bbd260
- **aiter**: `amd-aiter` 0.1.21.post2 (prebuilt `.co` from `aiter_meta/hsa/gfx950/mla/`)
- **dtype**: bf16 Q, bf16 KV
- **row width**: 576 elements (`kv_lora_rank` 512 + `qk_rope_head_dim` 64), 1152 bytes

## The threshold

For a 576-element bf16 row the cliff is exact and analytic:

```
2**32 // 1152 == (2**31 - 1) // 576 == 3,728,270
```

Row 3,728,269 is fully correct; row 3,728,270 straddles the boundary and comes
back **121 of 512 elements correct**; row 3,728,271 is entirely wrong. A
row-granular index truncation cannot produce a half-correct row — only
byte-granular address truncation can. The surviving prefix ends at the element
where the byte counter reaches `2**32` (128 elements into the row; the
~8-element shortfall is the last 16-byte vector load that would cross the
boundary).

## Reproducer

Pure AITER + PyTorch, no vLLM. Needs ~6 GiB of VRAM.

```python
"""AITER bf16 MLA decode wraps past a 4 GiB window from ptr_KV (gfx950)."""

import torch
from aiter.mla import mla_decode_fwd

ROW = 576   # kv_lora_rank 512 + qk_rope_head_dim 64
LORA = 512  # the kernel returns only the latent part
PAGE = 16
NKV = 64    # KV rows attended per query
CLIFF = 2**32 // (ROW * 2)  # == 3,728,270

torch.manual_seed(0)
dev = "cuda"
scale = 1.0 / (ROW**0.5)


def decode(kv, idx, q):
    """Sparse-MLA decode: present the buffer as [rows, page=1, nhead_kv=1, dim]."""
    heads = q.shape[1]
    out = torch.empty(1, heads, LORA, dtype=torch.bfloat16, device=dev)
    mla_decode_fwd(
        q,
        kv.view(-1, 1, 1, ROW),
        out,
        torch.tensor([0, 1], dtype=torch.int32, device=dev),
        torch.tensor([0, idx.numel()], dtype=torch.int32, device=dev),
        idx,
        torch.ones(1, dtype=torch.int32, device=dev),
        1,
        sm_scale=scale,
    )
    torch.cuda.synchronize()
    return out[0].float()


def relerr(kv, base_row, rows, q):
    """Max elementwise error vs an fp32 reference. ~0.003 is bf16 rounding."""
    idx = torch.arange(base_row, base_row + NKV, dtype=torch.int32, device=dev)
    out = decode(kv, idx, q)
    logits = (q[0].float() @ rows.float().T) * scale
    ref = torch.softmax(logits, dim=-1) @ rows.float()[:, :LORA]
    return float((out - ref).abs().max() / ref.abs().max().clamp_min(1e-6))


print(f"row = {ROW} elements = {ROW * 2} bytes (bf16)")
print(f"predicted cliff: 2**32 // {ROW * 2} = {CLIFF}\n")

kv = torch.zeros((270_000, PAGE, ROW), dtype=torch.bfloat16, device=dev)
rows = torch.randn(NKV, ROW, dtype=torch.bfloat16, device=dev)
q16 = torch.randn(1, 16, ROW, dtype=torch.bfloat16, device=dev)

print("A. sweep base row (tensor fixed at 4.32M rows, ~4.8 GiB, gqa_ratio=16)")
for base in (0, 1_000_000, 3_000_000, CLIFF - 100, CLIFF, CLIFF + 1, 4_000_000):
    kv.zero_()
    kv.view(-1, ROW)[base : base + NKV] = rows
    print(f"   base_row={base:>9}  byte_off={base * ROW * 2:>12}  "
          f"relerr={relerr(kv, base, rows, q16):.5f}")

# A single-row attend has softmax == 1.0, so the reference is the row itself and
# surviving elements can be counted. relerr is a max over elements and would
# hide a partial read.
print("\nA2. single-row boundary: how much of the straddling row survives?")
for r in range(CLIFF - 1, CLIFF + 2):
    row = torch.randn(1, ROW, dtype=torch.bfloat16, device=dev)
    kv.view(-1, ROW)[r] = row
    out = decode(kv, torch.tensor([r], dtype=torch.int32, device=dev), q16)
    ref = row.float()[:, :LORA]
    ok = (out - ref).abs().le(0.05 * ref.abs().clamp_min(1e-3)).all(0)
    cut = (2**32 - r * ROW * 2) // 2
    print(f"   row {r}: {int(ok.sum()):>3}/{LORA} elements correct"
          f"   (window ends {cut:>5} elements into the row)")

print("\nB. identical bytes, only ptr_KV moves")
BASE_ROW = 4_000_000
SHIFT_PAGES = 243_750  # * 16 = 3.9M rows, a whole number of pages
kv.zero_()
kv.view(-1, ROW)[BASE_ROW : BASE_ROW + NKV] = rows
print(f"   whole tensor   ptr_KV=kv         index={BASE_ROW:<9} "
      f"relerr={relerr(kv, BASE_ROW, rows, q16):.5f}")
shift_rows = SHIFT_PAGES * PAGE
print(f"   rebased view   ptr_KV=kv[{SHIFT_PAGES}:]  index={BASE_ROW - shift_rows:<9} "
      f"relerr={relerr(kv[SHIFT_PAGES:], BASE_ROW - shift_rows, rows, q16):.5f}")
print("   Same bytes, same physical location. Only ptr_KV differs.")

# Self-consistency rather than the fp32 reference: identical q and identical KV
# values, placed once below the 4 GiB mark and once above it. Each gqa_ratio
# selects a different kernel with its own numerics, but no kernel's answer may
# depend on where in the buffer the rows happen to live.
print("\nC. which gqa_ratio variants wrap? (same values, low vs high offset)")
q_all = torch.randn(1, 128, ROW, dtype=torch.bfloat16, device=dev)
for heads in (16, 32, 64, 128):
    q = q_all[:, :heads].contiguous()
    outs = []
    for base in (1000, 4_000_000):
        kv.zero_()
        kv.view(-1, ROW)[base : base + NKV] = rows
        idx = torch.arange(base, base + NKV, dtype=torch.int32, device=dev)
        outs.append(decode(kv, idx, q))
    low, high = outs
    diff = float((high - low).abs().max() / low.abs().max().clamp_min(1e-6))
    verdict = "OK" if diff < 0.05 else "WRAPS"
    print(f"   gqa_ratio={heads:>4}  reldiff(high vs low)={diff:>8.5f}  {verdict}")
```

### Output on the environment above

```
row = 576 elements = 1152 bytes (bf16)
predicted cliff: 2**32 // 1152 = 3728270

A. sweep base row (tensor fixed at 4.32M rows, ~4.8 GiB, gqa_ratio=16)
   base_row=        0  byte_off=           0  relerr=0.00297
   base_row=  1000000  byte_off=  1152000000  relerr=0.00297
   base_row=  3000000  byte_off=  3456000000  relerr=0.00297
   base_row=  3728170  byte_off=  4294851840  relerr=0.00297
   base_row=  3728270  byte_off=  4294967040  relerr=0.99651
   base_row=  3728271  byte_off=  4294968192  relerr=1.00000
   base_row=  4000000  byte_off=  4608000000  relerr=1.00000

A2. single-row boundary: how much of the straddling row survives?
   row 3728269: 512/512 elements correct   (window ends   704 elements into the row)
   row 3728270: 121/512 elements correct   (window ends   128 elements into the row)
   row 3728271:   0/512 elements correct   (window ends  -448 elements into the row)

B. identical bytes, only ptr_KV moves
   whole tensor   ptr_KV=kv         index=4000000   relerr=1.00000
   rebased view   ptr_KV=kv[243750:]  index=100000    relerr=0.00297
   Same bytes, same physical location. Only ptr_KV differs.

C. which gqa_ratio variants wrap? (same values, low vs high offset)
   gqa_ratio=  16  reldiff(high vs low)= 1.00000  WRAPS
   gqa_ratio=  32  reldiff(high vs low)= 1.00000  WRAPS
   gqa_ratio=  64  reldiff(high vs low)= 0.00000  OK
   gqa_ratio= 128  reldiff(high vs low)= 1.00000  WRAPS
```

## Why this is reachable in practice

A single request never accumulates 3.7M KV rows, so a conventional serving
configuration stays under the limit by a wide margin. What reaches it is a
**pool-global row index**.

vLLM's HiSparsekeeps the bulk of the sparse-MLA KV cache in pinned
host memory and points attention at a view over the **whole multi-layer KV
slab**, so the row indices it emits scale with the total block count rather than
with one layer's page table. Measured on GLM-5.3 (78 layers, TP4, MI350X):

| arm | KV tensor shape | flat rows | max row index emitted |
|---|---|---|---|
| HiSparse off | `(57773, 16, 576)` | 924,368 | **33** |
| HiSparse on | `(4643148, 16, 576)` | 74,290,368 | **7,541,967** |

Off-arm never comes within 4× of the limit. On-arm is 2× past it. End-to-end
that presented as GSM8K **0.000** with HiSparse on versus **0.905** off, with no
error of any kind anywhere in the stack.
